### PR TITLE
SOAP-3262 add new menu items

### PR DESCRIPTION
**Pull request by EkaterinaSorokina**

### CODE REVIEW

File: soapui/src/main/java/com/eviware/soapui/impl/wsdl/actions/testcase/WsdlTestCaseSoapUIActionGroup.java [hunk 1]
```diff
@@ -43,7 +43,7 @@ public WsdlTestCaseSoapUIActionGroup(String id, String name) {
             }
         }
 
-        if (toggleDisabledActionMapping != null) {
+        if (toggleDisabledActionMapping != null && modelItem!=null) {
```
Comment (IlyaAvdeev) [inline soapui/src/main/java/com/eviware/soapui/impl/wsdl/actions/testcase/WsdlTestCaseSoapUIActionGroup.java]: please use Ctrl+Alt+I

File: soapui/src/main/java/com/eviware/soapui/impl/wsdl/actions/teststep/WsdlTestStepSoapUIActionGroup.java [hunk 2]
```diff
@@ -80,7 +80,7 @@ public WsdlTestStepSoapUIActionGroup(String id, String name) {
             }
         }
 
-        if (toggleDisabledActionMapping != null) {
+        if (toggleDisabledActionMapping != null && modelItem!=null) {
```
Comment (IlyaAvdeev) [inline soapui/src/main/java/com/eviware/soapui/impl/wsdl/actions/teststep/WsdlTestStepSoapUIActionGroup.java]: see previous comment

File: soapui/src/main/java/com/eviware/soapui/impl/wsdl/actions/testsuite/WsdlTestSuiteSoapUIActionGroup.java [hunk 3]
```diff
@@ -43,7 +43,7 @@ public WsdlTestSuiteSoapUIActionGroup(String id, String name) {
             }
         }
 
-        if (toggleDisabledActionMapping != null) {
+        if (toggleDisabledActionMapping != null && modelItem!=null) {
```
Comment (IlyaAvdeev) [inline soapui/src/main/java/com/eviware/soapui/impl/wsdl/actions/testsuite/WsdlTestSuiteSoapUIActionGroup.java]: see previous comment

File: soapui/src/main/java/com/eviware/soapui/support/swing/FunctionalMenu.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+import com.eviware.soapui.support.action.support.DefaultActionMapping;
+import com.eviware.soapui.support.action.swing.ActionList;
+import com.eviware.soapui.support.action.swing.ActionListBuilder;
+import com.eviware.soapui.support.action.swing.ActionSupport;
+import com.eviware.soapui.support.action.swing.SwingActionDelegate;
+import javax.swing.*;
+import java.awt.*;
+
+
+public class FunctionalMenu {
```
Comment (IlyaAvdeev) [inline soapui/src/main/java/com/eviware/soapui/support/swing/FunctionalMenu.java]: I think MenuBuilderHelper is better naming for this class

File: soapui/src/main/java/com/eviware/soapui/support/swing/FunctionalMenu.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    public static JMenu getMenu(String name){
+        for (int i = 0; i<SoapUI.getMenuBar().getMenuCount(); i++){
```
Comment (IlyaAvdeev) [inline soapui/src/main/java/com/eviware/soapui/support/swing/FunctionalMenu.java]: Ctrl + Alt + I

File: soapui/src/main/java/com/eviware/soapui/support/swing/FunctionalMenu.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+            if (SoapUI.getMenuBar().getMenu(i).getText().equals(name)){
+                return SoapUI.getMenuBar().getMenu(i);
+            }
+        }
+        return null;
+    }
```
Comment (IlyaAvdeev) [inline soapui/src/main/java/com/eviware/soapui/support/swing/FunctionalMenu.java]: this function uses SoapUI.getMenuBar() at 3 places. One particular variable should be extracted before "for" loop.

File: soapui/src/main/java/com/eviware/soapui/support/swing/FunctionalMenu.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+            component.setEnabled(false);
+        }
+        return menu;
+    }
+
+    protected static ActionList buildActionsForActionGroup(String actionGroup) {
+        return ActionListBuilder.buildActions(actionGroup, null);
+    }
+
+    private static void disableMenuItems(String groupId) {
```
Comment (IlyaAvdeev) [inline soapui/src/main/java/com/eviware/soapui/support/swing/FunctionalMenu.java]: ooooops, disableMenuItems and enableMenuItems should be replaced with one function activateMenuItems (String groupId, boolean activate). See DRY principle https://ru.wikipedia.org/wiki/Don%E2%80%99t_repeat_yourself

File: soapui/src/main/java/com/eviware/soapui/support/swing/FunctionalMenu.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+    private static void enableMenuItems(String groupId) {
+        JMenu menu = getMenu(groupId);
+        for (Component component : menu.getMenuComponents()) {
+            disableEnableMenuSubItems(component, true);
+            component.setEnabled(true);
+        }
+    }
+
+    private static void disableEnableMenuSubItems(Component component, boolean bEnable) {
```
Comment (IlyaAvdeev) [inline soapui/src/main/java/com/eviware/soapui/support/swing/FunctionalMenu.java]: bad naming, activateMenuSubItems

File: soapui/src/main/java/com/eviware/soapui/support/swing/FunctionalMenu.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+            for (Component curComponent : ((JMenu) component).getMenuComponents()) {
+                curComponent.setEnabled(bEnable);
+            }
+        }
+    }
+
+    public static void buildMenu(ModelItem curModelItem, String menuName, SoapUITreeNode path) {
+        String[] groupsId = {SoapUI.STEP, SoapUI.CASE, SoapUI.SUITE, SoapUI.PROJECT};
+        if(curModelItem instanceof Workspace){
+            for (int i = 0; i<groupsId.length; i++) {
```
Comment (IlyaAvdeev) [inline soapui/src/main/java/com/eviware/soapui/support/swing/FunctionalMenu.java]: Ctrl+Alt+I

File: soapui/src/main/java/com/eviware/soapui/support/swing/FunctionalMenu.java [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+                disableMenuItems(groupsId[i]);
+            }
+        }else if (ModelSupport.isOneOf(curModelItem, WsdlTestSuite.class, WsdlTestCase.class,
+                WsdlTestStep.class, WsdlProject.class)) {
+            ActionList actionList = ActionListBuilder.buildActions(curModelItem);
+            JMenu curMenu = FunctionalMenu.getMenu(menuName);
+            curMenu.removeAll();
+            ActionSupport.addActions(actionList, curMenu);
+            for (int i = 0; !groupsId[i].equals(menuName); i++) {
```
Comment (IlyaAvdeev) [inline soapui/src/main/java/com/eviware/soapui/support/swing/FunctionalMenu.java]: is this correct? for integer i from 0 while false and at the end increment i.

File: soapui/src/main/java/com/eviware/soapui/support/swing/FunctionalMenu.java [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+                    buildMenu(curModelItem, SoapUI.PROJECT, null);
+                    break;
+                }
+                node = node.getParentTreeNode();
+                curModelItem = node.getModelItem();
+            }
+        }
+    }
+
+    public static String selectMenuName(ModelItem modelItem){
```
Comment (IlyaAvdeev) [inline soapui/src/main/java/com/eviware/soapui/support/swing/FunctionalMenu.java]: better naming is getMenuNameForModelItem

File: soapui/src/main/java/com/eviware/soapui/support/swing/FunctionalMenu.java (hunk 10 above)
Comment (divivitaly) [inline soapui/src/main/java/com/eviware/soapui/support/swing/FunctionalMenu.java]: Strange condition in the for loop... I can't understand it in 5 seconds...

File: soapui/src/main/java/com/eviware/soapui/ui/Navigator.java [hunk 12]
```diff
@@ -254,18 +253,18 @@ public void valueChanged(TreeSelectionEvent e) {
             Object obj = e.getPath().getLastPathComponent();
             if (obj instanceof SoapUITreeNode) {
                 SoapUITreeNode treeNode = (SoapUITreeNode) obj;
-
+                MenuBuilderHelper.buildMenu(treeNode.getModelItem(), MenuBuilderHelper.getMenuNameForModelItem(treeNode.getModelItem()), treeNode);
```
Comment (divivitaly) [inline soapui/src/main/java/com/eviware/soapui/ui/Navigator.java]: Maybe it better to reduce params to treeNode only?

File: soapui/src/main/java/com/eviware/soapui/support/swing/MenuBuilderHelper.java [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+            }
+        } else {
+            SoapUITreeNode node = path.getParentTreeNode();
+            curModelItem = node.getModelItem();
+            while (!(curModelItem instanceof Workspace)) {
+                if (curModelItem instanceof WsdlTestCase) {
+                    buildMenu(curModelItem, SoapUI.CASE, null);
+                    break;
+                }
+                if (curModelItem instanceof WsdlProject) {
```
Comment (divivitaly) [inline soapui/src/main/java/com/eviware/soapui/support/swing/MenuBuilderHelper.java]: Using if...else...if is more clearly way to understand.

**Comment by divivitaly** [comment]
Now I'm happy :-)

**Pull request merged**